### PR TITLE
Stream output when calling conda cli

### LINKED ISSUE #9412: Provide a way to run `conda run` w/o buffering stdout (and output in general)
Currently `conda run` buffers everything going to stdout (I didn't test stderr). It would be great to be able to get streaming results from the running command rather than having to wait until the application exits.

### LINKED ISSUE #9572: Conda run fails when running a python file containing `input` (reading from stdin)
## Current Behavior

* Create a python file as follows:
```python
import time
x = input("Hello")
print(x)
```
* Assume you have a conda environment named `env01`
* Conda version 4.7.12 on Mac
* Run in terminal as follows `conda run -n=env01 python two.py`
* The output isn't displayed due to https://github.com/conda/conda/issues/9412
    * Hit `Enter` and the output is displayed.
* Thats fine.
* Update to conda 4.8.1
* Run in terminal as follows `conda run -n=env01 python two.py`
```
ERROR conda.cli.main_run:execute(30): Subprocess for 'conda run ['python', 'two.py']' command failed.  (See above for error)
Hello
Traceback (most recent call last):
  File "two.py", line 2, in <module>
    x = input("Hello")
EOFError: EOF when reading a line
```
 
### Steps to Reproduce

```

```

## Expected Behavior

Running python files with `input` should work when using `conda run`.
It used to work in conda version 4.7.12 and doesn't anymore in 4.8.1

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
     active environment : None
            shell level : 0
       user config file : /Users/donjayamanne/.condarc
 populated config files : /Users/donjayamanne/.condarc
          conda version : 4.8.1
    conda-build version : not installed
         python version : 3.7.4.final.0
       virtual packages : __osx=10.15.2
       base environment : /Users/donjayamanne/miniconda3  (writable)
           channel URLs : https://repo.anaconda.com/pkgs/main/osx-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/osx-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /Users/donjayamanne/miniconda3/pkgs
                          /Users/donjayamanne/.conda/pkgs
       envs directories : /Users/donjayamanne/miniconda3/envs
                          /Users/donjayamanne/.conda/envs
               platform : osx-64
             user-agent : conda/4.8.1 requests/2.22.0 CPython/3.7.4 Darwin/19.2.0 OSX/10.15.2
                UID:GID : 501:20
             netrc file : None
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
==> /Users/donjayamanne/.condarc <==
auto_activate_base: False
```
</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
# packages in environment at /Users/donjayamanne/miniconda3:
#
# Name                    Version                   Build  Channel
asn1crypto                1.2.0                    py37_0    defaults
ca-certificates           2019.11.27                    0    defaults
certifi                   2019.11.28               py37_0    defaults
cffi                      1.13.2           py37hb5b8e2f_0    defaults
chardet                   3.0.4                 py37_1003    defaults
conda                     4.8.1                    py37_0    defaults
conda-package-handling    1.6.0            py37h1de35cc_0    defaults
cryptography              2.8              py37ha12b0ac_0    defaults
idna                      2.8                      py37_0    defaults
libcxx                    4.0.1                hcfea43d_1    defaults
libcxxabi                 4.0.1                hcfea43d_1    defaults
libedit                   3.1.20181209         hb402a30_0    defaults
libffi                    3.2.1                h475c297_4    defaults
ncurses                   6.1                  h0a44026_1    defaults
openssl                   1.1.1d               h1de35cc_3    defaults
pip                       19.3.1                   py37_0    defaults
pycosat                   0.6.3            py37h1de35cc_0    defaults
pycparser                 2.19                     py37_0    defaults
pyopenssl                 19.1.0                   py37_0    defaults
pysocks                   1.7.1                    py37_0    defaults
python                    3.7.4                h359304d_1    defaults
python.app                2                        py37_9    defaults
readline                  7.0                  h1de35cc_5    defaults
requests                  2.22.0                   py37_1    defaults
ruamel_yaml               0.15.87          py37h1de35cc_0    defaults
setuptools                44.0.0                   py37_0    defaults
six                       1.13.0                   py37_0    defaults
sqlite                    3.30.1               ha441bb4_0    defaults
tk                        8.6.8                ha441bb4_0    defaults
tqdm                      4.40.2                     py_0    defaults
urllib3                   1.25.7                   py37_0    defaults
wheel                     0.33.6                   py37_0    defaults
xz                        5.2.4                h1de35cc_4    defaults
yaml                      0.1.7                hc338f04_2    defaults
zlib                      1.2.11               h1de35cc_3    defaults
```
</p></details>

### PR DESCRIPTION
**Pull request by DonJayamanne**
When calling the conda cli this will:
* Stream output into stdout & stderr
* Stdin is passed into the subprocess

Basically, when using the CLI we won't be capturing the output, we will write output directly into stdout and stderr. If there's an error we throw the error as per usual and error info has already been written out to the corresponding streams.

Resolves #9572
Resolves #9412

### CODE REVIEW

**Comment by msarahan** [comment]
Thanks @DonJayamanne! It seems like this is still a bit unhappy with the way that subprocess deals with redirected in/out.  Can you help us understand how much time you have to spend on this?

@cla-bot check

**Comment by DonJayamanne** [comment]
>  like this is still a bit unhappy with the way that subprocess deals with

Fixed a bug with existing code for handling of `stdin`.
Should be ok now.

**Comment by DonJayamanne** [comment]
> Can you help us understand how much time you have to spend on this?

Only a couple of hours.
Didn't run any of the tests locally. Looks like it's still failing. Will look into them today.

**Comment by DonJayamanne** [comment]
@msarahan 
I believe some of the tests are existing failures. I will need to have a closer look.

**Comment by DonJayamanne** [comment]
> Can you help us understand how much time you have to spend on this?

Apologies, misread the question. 
@msarahan I can spend a few hours on this. How can I help?

**Comment by msarahan** [comment]
Any of this pseudo-file stuff is new, I think:

```
Traceback (most recent call last):
  File "/home/circleci/conda/tests/test_create.py", line 1740, in test_conda_pip_interop_dependency_satisfied_by_pip
    pip_o, pip_e, _ = run_command(Commands.RUN, prefix, "--dev", "python", "-m", "pip", "install", "itsdangerous")
  File "/home/circleci/conda/tests/test_create.py", line 281, in run_command
    result = do_call(args, p)
  File "/home/circleci/conda/conda/cli/conda_argparse.py", line 82, in do_call
    return getattr(module, func_name)(args, parser)
  File "/home/circleci/conda/conda/cli/main_run.py", line 29, in execute
    stdout=sys.stdout, stderr=sys.stderr, stdin=sys.stdin)
  File "/home/circleci/conda/conda/gateways/subprocess.py", line 73, in subprocess_call
    stderr=stderr, env=env)
  File "/opt/conda/lib/python3.7/subprocess.py", line 728, in __init__
    errread, errwrite) = self._get_handles(stdin, stdout, stderr)
  File "/opt/conda/lib/python3.7/subprocess.py", line 1347, in _get_handles
    p2cread = stdin.fileno()
  File "/opt/conda/lib/python3.7/site-packages/_pytest/capture.py", line 731, in fileno
    raise UnsupportedOperation("redirected stdin is pseudofile, has no fileno()")
io.UnsupportedOperation: redirected stdin is pseudofile, has no fileno()
```

I was asking how much help you needed in a roundabout, awkward way.  If you're stuck, please let us know and we can try to help.

**Comment by DonJayamanne** [comment]
Let me know whether mocking sys.stdin was ok.
I was planning on doing `self.monkeypatch.setattr('sys.stdin', sys.stdout)`, but i would look confusing, hence used StringIO with the `fileno` attribute.

**Comment by DonJayamanne** [comment]
@msarahan if it doesn't work please could you let me know what I could/need to do to get this through.
Thanks for your help.

**Comment by msarahan** [comment]
Looks fine conceptually to me.  Tests still failing - but now in a new way.  It looks like None is getting returned somehow:

```
Traceback (most recent call last):
  File "/home/circleci/conda/tests/test_create.py", line 1445, in test_conda_run
    assert output == 'hello' + os.linesep
AssertionError: assert None == 'hello\n'
  -None
  +'hello\n'
```

**Comment by DonJayamanne** [comment]
Yes, that's because the output is no longer piped and returned.
Chris, why do we need the output returned in such a manner is it for testing purposes?
When looking at the code we print the output back again. Hence writing to the output stream should suffice and change the tested to deal with capturing text written to stdout and stderr.

Let me know what you think

**Comment by shenker** [comment]
Hi all, any progress on this? I'd find this super useful!

**Comment by DonJayamanne** [comment]
Oh I'm so sorry for forgetting this. Will get the tests fixed asap

**Comment by mjurkus** [comment]
Any updates on this?

**Comment by DonJayamanne** [comment]
@msarahan 
Any idea why the appveyor tests are failing?

File: conda/activate.py [hunk 1]
```diff
@@ -641,7 +641,7 @@ def _build_activate_shell_custom(self, export_vars):
         # The signature of this method may change in the future.
         if on_win:
             import ctypes
-            export_vars.update({"PYTHONIOENCODING": ctypes.cdll.kernel32.GetACP(),})
+            export_vars.update({"PYTHONIOENCODING": ctypes.cdll.kernel32.GetACP(), })
```
Comment (DonJayamanne) [inline conda/activate.py:644]: I didn't touch this file, however flake8 was complaining about this file in my PR, hence the change to this file.

Comment (angloyna) [reply conda/activate.py:644]: this line looks like it's been removed 👍 needs a merge conflict fix, though

**Comment by zhc7** [comment]
I've got the same issue as #9572 , any updates?

**Comment by angloyna** [comment]
@zhc7 IIRC, i think @DonJayamanne has a merge conflict unrelated to his code change that he should be able to fix and get this in 👍

**Comment by shenker** [comment]
@DonJayamanne would be great to get this in, if you've got a chance. thanks for your work on this!

**Comment by DonJayamanne** [comment]
@msarahan 
would appreciate some help in getting the following tests fixed
```
FAILED tests/test_cli.py::TestRun::test_run_returns_int - Exception: Got stde...
FAILED tests/test_cli.py::TestRun::test_run_returns_nonzero_errorlevel - Exce...
FAILED tests/test_cli.py::TestRun::test_run_returns_zero_errorlevel - Excepti...
```
Basically its an issue with output.
Not sure whats going on here. The CLI doesn't dump any output, but it seems to capture output from the tests & that interferers with the test execution.

**Comment by travishathaway** [comment]
This pull request has become stale and is therefore being closed. Please re-open if you would like to continue development on this.

**Pull request closed without merging**